client/src/index.js

```diff
@@ -1,41 +1,219 @@
-import React from "react";
+import React, { Component } from "react";
```

Comment:
This component really hard to understand and too long, usually the index file rendering hall the app will be simple and does not has to much code so I refactored this component to four components I will add four of them to this comment with their names and structures. 

1- This index.js file has only the down code:

```import React from "react";
import ReactDOM from "react-dom";
import Routes from './Routes';
import "semantic-ui-css/semantic.min.css";
import "./index.css";

ReactDOM.render(<Routes />, document.getElementById("root"));
```

2- Route.js component should only care about handling routes no more others scenarios, so the code for it down, should be named Routes.js as imported on index.js file.

```import React from "react";
import { BrowserRouter as Router, Route } from "react-router-dom";
import Home from "./components/Home";
import About from "./components/About";
import Status from "./components/Status";
import Login from "./components/Login";
import ApplicantRegister from "./components/ApplicantRegister";
import NewOpportunityForm from "./components/NewOpportunityForm";
import MainRegister from "./components/MainRegister";
import ApplicantProfile from "./components/ApplicantProfile";
import NavBar from "./Navbar";
import { Container } from "semantic-ui-react";

export default () => {
  return (
    <Router>
      <NavBar />
      <Container>
      <Route path="/" exact component={Home} />
      <Route path="/about" exact component={About} />
      <Route path="/status" exact component={Status} />
      <Route path="/login" exact component={Login} />
      <Route path="/applicant-profile" exact component={ApplicantProfile} />
      <Route path="/create-opportunity" exact component={NewOpportunityForm} />
      <Route path="/main-register" exact component={MainRegister} />
      <Route path="/applicant-register" exact component={ApplicantRegister} />
      </Container>
    </Router>
  );
};
```

3- Navbar.js again try to make small components because of that I made these two components first one is Navbar and the second is SideBarMenu the down code for Navbar.js

```import React, { Component } from "react";
import "semantic-ui-css/semantic.min.css";
import "./index.css";
import { Link } from "react-router-dom";
import { Menu, Icon, Button } from "semantic-ui-react";
import SideBarMenu from "./SideBarMenu";

export default class NavBar extends Component {
  state = {
    visible: false,
    activeItem: "home"
  };

  handleItemClick = (e, { name }) => this.setState({ activeItem: name });

  logout = event => {
    event.preventDefault();
    localStorage.removeItem("token");
    document.location.reload();
  };

  handleHideClick = () => this.setState({ visible: false });
  handleShowClick = () => this.setState({ visible: true });
  handleSidebarHide = () => this.setState({ visible: false });

  handleItemClick = (e, { name }) => this.setState({ activeItem: name });

  render() {
    const { visible, activeItem } = this.state;
    const { handleSidebarHide, handleItemClick, logout } = this;
    return (
      <React.Fragment>
        <Menu inverted style={{ 'border-radius': 'inherit' }}>
          <Menu.Item>
            <Button disabled={visible} onClick={this.handleShowClick}>
              <Icon name="list layout"></Icon>
            </Button>
          </Menu.Item>
          <Menu.Item
            name="Logout"
            active={activeItem === "Logout"}
            onClick={event => {
              this.logout(event);
              this.handleItemClick(event);
            }}
            as={Link}
            to="/logout"
            position="right"
          >
            <Button inverted basic>
              {activeItem}
            </Button>
          </Menu.Item>
          {localStorage.getItem("token") ? (
            <Menu.Item
              name="Logout"
              active={activeItem === "Logout"}
              onClick={event => {
                this.logout(event);
                this.handleItemClick(event);
              }}
              as={Link}
              position="right"
              to="/logout"
            >
              <Button primary> Logout</Button>
            </Menu.Item>
          ) : (
            <Menu.Item
              name="Sign Up"
              active={activeItem === "Sign Up"}
              onClick={this.handleItemClick}
              as={Link}
              to="/login"
              position="right"
            >
              {" "}
              <Button primary>Sign in</Button>
            </Menu.Item>
          )}
        </Menu>
        <SideBarMenu
          visible={visible}
          handleSidebarHide={handleSidebarHide}
          activeItem={activeItem}
          handleItemClick={handleItemClick}
          logout={logout}
        />
      </React.Fragment>
    );
  }
}
```

4- SideBarMenu.js the last component should care only about sidebar contents.

```import React from "react";
import "semantic-ui-css/semantic.min.css";
import "./index.css";
import { Link } from "react-router-dom";
import { Sidebar, Menu, Icon, Button } from "semantic-ui-react";

export default ({ visible, handleSidebarHide, activeItem, handleItemClick, logout }) => {
  return (
    <Sidebar
      as={Menu}
      animation="overlay"
      icon="labeled"
      inverted
      onHide={handleSidebarHide}
      vertical
      visible={visible}
      width="thin"
      style={{
        top: '48px',
      }}
    >
      <Menu.Item
        name="Home"
        onClick={handleItemClick}
        as={Link}
        active={activeItem === "Home"}
        to="/"
      >
        <Icon name="home"></Icon>
      </Menu.Item>
      <Menu.Item
        name="About"
        onClick={handleItemClick}
        active={activeItem === "About"}
        as={Link}
        to="/about"
      >
        <Icon name="adn"></Icon>
      </Menu.Item>
      <Menu.Item
        name="Status"
        onClick={handleItemClick}
        as={Link}
        active={activeItem === "Status"}
        to="/status"
      >
        <Icon name="star outline"></Icon>
      </Menu.Item>
      <Menu.Item
        name="Create Opportunity"
        onClick={handleItemClick}
        as={Link}
        active={activeItem === "Create Opportunity"}
        to="/create-opportunity"
      >
        <Icon name="star outline"></Icon>
      </Menu.Item>
      <Menu.Item
        name="Applicant Profile"
        onClick={handleItemClick}
        as={Link}
        active={activeItem === "Applicant Profile"}
        to="/applicant-profile"
      />
      {localStorage.getItem("token") ? (
        <Menu.Item
          name="Logout"
          active={activeItem === "Logout"}
          onClick={event => { logout(event); handleItemClick(event) }}
          as={Link}
          position="right"
          to="/logout"
        >
          <Icon name="log out"></Icon>

          <Button primary> Logout</Button>
        </Menu.Item>
      ) : (
        <Menu.Menu position="right">
          <Menu.Item
            name="Sign In"
            active={activeItem === "Sign In"}
            onClick={handleItemClick}
            as={Link}
            to="/login"
          >
            <Icon name="sign-in"></Icon>{" "}
          </Menu.Item>
          <Menu.Item
            name="Main register"
            active={activeItem === "Main register"}
            onClick={handleItemClick}
            as={Link}
            to="/main-register"
          >
            <Icon name="signup"></Icon>
          </Menu.Item>
        </Menu.Menu>
      )}
    </Sidebar>
  );
};
````

### Note:
- If you notice I used React.Fragment it is similar to div the only different div will reserve a space to itself and  React.Fragment will not use it only to wrap the contents in React render function.

- Also, I update onClick to call two functions correctly instead of deplicat it on the same element,
e.g ```onClick={event => { functionOne(event); functionTow(event) }}```

Reply:
done it is amazing you spend a lot of time  refactoring it :) thanks very much